src/wallet/rpcwallet.cpp

```diff
@@ -3150,6 +3150,55 @@ struct StakePeriodRange_T {
 
 typedef std::vector<StakePeriodRange_T> vStakePeriodRange_T;
 
+// Check if we have a Tx that can be counted in staking report
+bool IsTxCountedAsStaked(const CWalletTx* tx)
+{
+    // orphan block or immature
+    if ((!tx->GetDepthInMainChain()) || (tx->GetBlocksToMaturity() > 0))
+        return false;
+
+    // abandoned transactions
+    if (tx->isAbandoned())
+        return false;
+
+    // transaction other than POS block
+    return tx->IsCoinStake();
+}
+
+// Get the amount for a staked tx used in staking report
+CAmount GetTxStakeAmount(const CWalletTx* tx)
+{
+    // use the cached amount if available
+    if ((tx->fCreditCached || tx->fColdStakingCreditCached) && (tx->fDebitCached || tx->fColdStakingDebitCached))
+        return tx->nCreditCached + tx->nColdStakingCreditCached - tx->nDebitCached - tx->nColdStakingDebitCached;
+    // Check for cold staking
+    else if (tx->vout[1].scriptPubKey.IsColdStaking())
+        return tx->GetCredit(pwalletMain->IsMine(tx->vout[1])) - tx->GetDebit(pwalletMain->IsMine(tx->vout[1]));
+
+    return tx->GetCredit(ISMINE_SPENDABLE) + tx->GetCredit(ISMINE_STAKABLE) - tx->GetDebit(ISMINE_SPENDABLE) - tx->GetDebit(ISMINE_STAKABLE);
+}
+
+// Gets timestamp for first stake
+// Returns -1 (Zero) if has not staked yet
+int64_t GetFirstStakeTime()
+{
+    // Need a pointer for the tx
+    const CWalletTx* tx;
+
+    // scan the entire wallet transactions
+    for (auto it = pwalletMain->wtxOrdered.begin(); it != pwalletMain->wtxOrdered.end(); ++it)
```

Comment:
If wtxOrdered is not changing length during this loop, you could cache the end iterator to save it being calculated each time the loop iterates. I'm not sure on the exact syntax here, in javascript you can do it directly into the loop declaration like this;

```
for(let i=0, l=my_array.length(); i < l; i++) {
    console.log(my_array[i]);
}
```
I don't think there's an inline way to do it in C++, so you'd just cache the iterator before you start the loop something like this;
```
std::vector<uint32_t>::const_iterator end = pwalletMain->wtxOrdered.end();
or (auto it = pwalletMain->wtxOrdered.begin(); it != end; ++it)
{
    //do stuff
}
```
It's probably not hugely relevant here since we're hoping the first item is the one we're looking for, but its a neat trick that can optimise loops especially when you're iterating through thousands of items like we can be doing when dealing with wallet transactions and especially long lived staking addresses. Keep it in mind if it's useful somewhere else :)